agent/xds/clusters.go

```diff
@@ -631,11 +633,51 @@ func (s *Server) makeGatewayCluster(snap *proxycfg.ConfigSnapshot, opts gatewayC
 	}
 	cluster.ClusterDiscoveryType = &discoveryType
 
-	endpoints := make([]envoyendpoint.LbEndpoint, 0, len(opts.hostnameEndpoints))
+	endpoints := make([]envoyendpoint.LbEndpoint, 0, 1)
+	uniqueHostnames := make(map[string]bool)
 
-	for _, e := range opts.hostnameEndpoints {
-		endpoints = append(endpoints, makeLbEndpoint(e, opts.isRemote, opts.onlyPassing))
+	var (
+		hostname string
+		idx      int
+	)
+	for i, e := range opts.hostnameEndpoints {
+		addr, port := e.BestAddress(opts.isRemote)
+		uniqueHostnames[addr] = true
+
+		health, weight := calculateEndpointHealthAndWeight(e, opts.onlyPassing)
+		if health == envoycore.HealthStatus_UNHEALTHY {
+			continue
+		}
+
+		if len(endpoints) == 0 {
+			endpoints = append(endpoints, makeLbEndpoint(addr, port, health, weight))
+
+			hostname = addr
+			idx = i
+		}
+	}
+
+	dc := opts.hostnameEndpoints[idx].Node.Datacenter
+	service := opts.hostnameEndpoints[idx].Service.CompoundServiceName()
+
+	loggerName := logging.TerminatingGateway
+	if snap.Kind == structs.ServiceKindMeshGateway {
+		loggerName = logging.MeshGateway
+	}
+
+	if len(endpoints) == 0 {
+		s.Logger.Named(loggerName).
+			Warn("service does not contain any healthy instances, skipping Envoy cluster creation",
+				"dc", dc, "service", service.String())
+
+		return nil
+	}
+	if len(uniqueHostnames) > 1 {
+		s.Logger.Named(loggerName).
+			Warn(fmt.Sprintf("service contains instances with more than one unique hostname; only %q be resolved by Envoy.", hostname),
```

Comment:
```suggestion
			Warn(fmt.Sprintf("service contains instances with more than one unique hostname; only %q be resolved by Envoy", hostname),
```

smallest of nitpicks, but with a period log messages look like so: `resolved by Envoy.: dc=dc1 service=hostname`

Reply:
fixed!